Change XML saving for maze to fix infrequent crash

mishmish66 opened the pull request:
# Description

I changed the way that temporary xmls are stored for the maze envs. I have used this change for training and it improves stability, without it my training would crash sometimes, usually after about 1m env steps. With this change it seems to be stable.

## Type of change

Please delete options that are not relevant.

- [x] Bug fix (non-breaking change which fixes an issue)

# Checklist:

- [x] I have run the [`pre-commit` checks](https://pre-commit.com/) with `pre-commit run --all-files` (see `CONTRIBUTING.md` instructions to set it up)
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation N/A
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

Kallinteris-Andreas commented:
Can you please elaborate on why the crashing was happening prior to this change?

mishmish66 commented:
Honestly it's a little unclear to me, the crash threw a XML parsing error giving a line and reason, but obviously the XML is supposed to be working. My guess is either the XML tool we are using to write to the XML file does not flush the buffers correctly internally, and since here we use the open() context manager it flushes as needed, or it's because there is a collision since I'm using a lot of parallelism and this is using just timestamping. With these changes the directories are unique regardless of time, and also the file is flushed fully.

mishmish66 commented:
@Kallinteris-Andreas is there anything else I should do to get this fix in?

Kallinteris-Andreas commented on tests/envs/maze/test_point_maze.py:
```diff
@@ -39,3 +39,23 @@ def test_goal_cell():
     np.testing.assert_almost_equal(desired_goal, obs["desired_goal"], decimal=4)
+
+
+def test_multiprocessing_not_crash():
```
The Problem with this kind of nondeterministic testing, is that I can verify it working or not 

I can not get it to fail with the version of maze

If it can not prove that a bug has been fixed, it should not be added

mishmish66 replied:
let me see if I can make a test that fails it, I think the actual problem is that when using Tianshou RL it recreates the env a bunch of times, so what I should be doing to test is creating a buttload of envs.

Kallinteris-Andreas reviewed (commented):
In principle, those changes are an improvement over the hacky solution with timestamping XML files I had come up with.

To elaborate a bit, the likely cause of the previous version crashing was because of possible timestamp aliasing, which this PR fixes by using `/tmp/` files

Other than the testing issue, it is ready for merge

Kallinteris-Andreas commented:
@mishmish66 any progress on finding the test?

Note: I am willing to merge without a test, since it is hard to assert concurrency behavior

mishmish66 commented:
No, I couldn't replicate the errors in a test which is weird. I was seeing this on a node in a cluster so I guess testing it locally is slightly different, but in any case I guess it isn't a very significant issue if it only happens on strangely configured cluster nodes.

Kallinteris-Andreas commented:
Just remove `test_multiprocessing_not_crash` and I can merge this, it is still an improvement over the existing implamention

Kallinteris-Andreas commented:
@mishmish66 rebase and remove the test, and I can merge

Kallinteris-Andreas commented:
@mishmish66 please remove tests and rebase

mishmish66 committed:
commented code

mishmish66 committed:
got rid of unused time import

mishmish66 committed:
made xml writing stuff for maze a bit more robust

mishmish66 committed:
added cleanup call to prevent tmp_dir from warning when garbage collecting

mishmish66 committed:
added tmp_dir stuff to maze_v3 as well

mishmish66 commented:
@Kallinteris-Andreas I rebased and fixed a warning that was thrown from the env on cleanup that was causing the tests to fail. I also added the change to maze-v3 envs, let me know if that is the wrong thing to do and I'll go remove that!

Thanks for your patience!

Kallinteris-Andreas commented on gymnasium_robotics/envs/maze/maze.py:
```diff
@@ -308,3 +316,7 @@ def compute_truncated(
     def update_target_site_pos(self, pos):
         raise NotImplementedError
+
+    def __del__(self):
```
Is this actually required? does it now cleanup in `/tmp` without it?

Kallinteris-Andreas reviewed (changes requested):
Sorry for late response, 

please run `pre-commit`

mishmish66 commented:
Hey sorry I never got back to this! I completely forgot, glad to see you got around to getting the fix in! Thanks for all your work maintaining gymnasium-robotics :)
